File: DSA/Tree/Code/DiameterOfTree.py
# Diameter of Tree in Python

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class DiameterOfTree:
    @staticmethod
    def get_height(root):
        if root is None:
            return 0
        
        return max(DiameterOfTree.get_height(root.left),
                   DiameterOfTree.get_height(root.right)) + 1
    
    @staticmethod
    def get_diameter_naive(root):
        """Naive approach: O(n²)"""
        if root is None:
            return 0
        
        left_height = DiameterOfTree.get_height(root.left)
        right_height = DiameterOfTree.get_height(root.right)
        diameter_through_root = left_height + right_height + 1
        
        left_diameter = DiameterOfTree.get_diameter_naive(root.left)
        right_diameter = DiameterOfTree.get_diameter_naive(root.right)
        
        return max(diameter_through_root, left_diameter, right_diameter)
    
    @staticmethod
    def get_diameter_optimized(root, diameter_list):
        """Optimized approach: O(n)"""
        if root is None:
            return 0
        
        left_height = DiameterOfTree.get_diameter_optimized(root.left, diameter_list)
        right_height = DiameterOfTree.get_diameter_optimized(root.right, diameter_list)
        
        current_diameter = left_height + right_height + 1
        diameter_list[0] = max(diameter_list[0], current_diameter)
        
        return max(left_height, right_height) + 1

# Test
print("=== Diameter of Tree (Python) ===\n")

# Create tree
root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)
root.right.left = Node(6)
root.right.right = Node(7)

print("Tree structure:")
print("         1")
print("        / \\")
print("       2   3")
print("      / \\ / \\")
print("     4  5 6  7\n")

print(f"Diameter (Naive approach): {DiameterOfTree.get_diameter_naive(root)}")

diameter_list = [0]
DiameterOfTree.get_diameter_optimized(root, diameter_list)
print(f"Diameter (Optimized approach): {diameter_list[0]}\n")

print("=== Diameter Definition ===")
print("Diameter: Longest path between any two nodes in tree")
print("Includes: Number of edges in the path\n")

print("Possible cases for diameter:")
print("1. Path passes through root")
print("   diameter = leftHeight + rightHeight + 1")
print("2. Path entirely in left subtree")
print("3. Path entirely in right subtree\n")

print("=== Complexity Analysis ===")
print("Naive Approach:")
print("  Time Complexity:  O(n²) - getHeight called for each node")
print("  Space Complexity: O(h) - recursion stack\n")

print("Optimized Approach:")
print("  Time Complexity:  O(n) - single pass")
print("  Space Complexity: O(h) - recursion stack")

File: DSA/Tree/Code/DiameterOfTree.cpp
#include <iostream>
#include <algorithm>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    
    Node(int val) : data(val), left(NULL), right(NULL) {}
};

class DiameterOfTree {
private:
    static int getHeight(Node* root) {
        if (root == NULL) {
            return 0;
        }
        
        return max(getHeight(root->left), getHeight(root->right)) + 1;
    }
    
public:
    // Naive approach: O(n²)
    static int getDiameterNaive(Node* root) {
        if (root == NULL) {
            return 0;
        }
        
        int leftHeight = getHeight(root->left);
        int rightHeight = getHeight(root->right);
        int diameterThroughRoot = leftHeight + rightHeight + 1;
        
        int leftDiameter = getDiameterNaive(root->left);
        int rightDiameter = getDiameterNaive(root->right);
        
        return max({diameterThroughRoot, leftDiameter, rightDiameter});
    }
    
    // Optimized approach: O(n)
    static int getDiameterOptimized(Node* root, int& diameter) {
        if (root == NULL) {
            return 0;
        }
        
        int leftHeight = getDiameterOptimized(root->left, diameter);
        int rightHeight = getDiameterOptimized(root->right, diameter);
        
        diameter = max(diameter, leftHeight + rightHeight + 1);
        
        return max(leftHeight, rightHeight) + 1;
    }
};

int main() {
    cout << "=== Diameter of Tree (C++) ===\n\n";
    
    // Create tree
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->left = new Node(6);
    root->right->right = new Node(7);
    
    cout << "Tree structure:\n";
    cout << "         1\n";
    cout << "        / \\\n";
    cout << "       2   3\n";
    cout << "      / \\ / \\\n";
    cout << "     4  5 6  7\n\n";
    
    cout << "Diameter (Naive approach): " << DiameterOfTree::getDiameterNaive(root) << "\n";
    
    int diameter = 0;
    DiameterOfTree::getDiameterOptimized(root, diameter);
    cout << "Diameter (Optimized approach): " << diameter << "\n\n";
    
    cout << "=== Diameter Definition ===\n";
    cout << "Diameter: Longest path between any two nodes in tree\n";
    cout << "Includes: Number of edges in the path\n\n";
    
    cout << "Possible cases for diameter:\n";
    cout << "1. Path passes through root\n";
    cout << "   diameter = leftHeight + rightHeight + 1\n";
    cout << "2. Path entirely in left subtree\n";
    cout << "3. Path entirely in right subtree\n\n";
    
    cout << "=== Complexity Analysis ===\n";
    cout << "Naive Approach:\n";
    cout << "  Time Complexity:  O(n²) - getHeight called for each node\n";
    cout << "  Space Complexity: O(h) - recursion stack\n\n";
    
    cout << "Optimized Approach:\n";
    cout << "  Time Complexity:  O(n) - single pass\n";
    cout << "  Space Complexity: O(h) - recursion stack\n";
    
    return 0;
}
